src/adapter/src/rbac.rs

```diff
@@ -0,0 +1,337 @@
+}
+
+impl fmt::Display for UnauthorizedError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "permission denied to {}", self.action)
+    }
+}
+
+impl UnauthorizedError {
+    pub fn superuser(action: String) -> UnauthorizedError {
+        UnauthorizedError {
+            action,
+            kind: UnauthorizedErrorKind::Superuser,
+        }
+    }
+
+    pub fn attribute(action: String, attribute: Attribute) -> UnauthorizedError {
+        UnauthorizedError {
+            action,
+            kind: UnauthorizedErrorKind::Attribute(attribute),
+        }
+    }
+
+    pub fn privilege(action: String, reason: String) -> UnauthorizedError {
+        UnauthorizedError {
+            action,
+            kind: UnauthorizedErrorKind::Unstructured { reason },
+        }
+    }
+
+    pub fn detail(&self) -> Option<String> {
+        match &self.kind {
+            UnauthorizedErrorKind::Superuser => {
+                Some(format!("You must be a superuser to {}", self.action))
+            }
+            UnauthorizedErrorKind::Attribute(attribute) => Some(format!(
+                "You must have the {} attribute to {}",
+                attribute.to_string(),
+                self.action
+            )),
+            UnauthorizedErrorKind::Unstructured { reason } => {
+                Some(format!("{} to {}", reason, self.action))
+            }
+        }
+    }
+}
+
+impl Error for UnauthorizedError {}
+
+/// Checks if a session is authorized to execute a command. If not, an error is returned.
+///
+/// Note: The session and role ID are stored in the command itself.
+pub fn check_command(catalog: &Catalog, cmd: &Command) -> Result<(), AdapterError> {
+    if !catalog.system_config().enable_rbac_checks() {
+        return Ok(());
+    }
+
+    let role = if !matches!(cmd, Command::Startup { .. }) {
+        if let Some(session) = cmd.session() {
+            let conn_catalog = catalog.for_session(session);
+            let role_id = session.role_id();
+            let role = conn_catalog.try_get_role(role_id);
+            if role.is_none() {
+                // PostgreSQL allows users that have their role dropped to perform some actions,
+                // such as `SET ROLE` and certain `SELECT` queries. We haven't implemented
+                // `SET ROLE` and feel it's safer to force to user to re-authenticate if their
+                // role is dropped.
+                return Err(AdapterError::ConcurrentRoleDrop(role_id.clone()));
+            }
+            role
+        } else {
+            None
+        }
+    } else {
+        None
+    };
+
+    match cmd {
+        Command::DumpCatalog { session, .. } => {
+            if session.is_superuser() {
+                Ok(())
+            } else {
+                Err(AdapterError::Unauthorized(UnauthorizedError::superuser(
+                    "dump catalog".into(),
+                )))
+            }
+        }
+        Command::Execute {
```

Comment:
Does the logic here need to exist? It's possible it was already doing the checking of the inner portal correctly. Can you add a test for `PREPARE; EXECUTE` (I'd like that no matter what)? Then, if you remove the logic here, do the tests still correctly flag it?

Reply:
You can't prepare DDL statements unfortunately, so there's nothing to test for this PR because only DDL statements are affected by attributes.

https://www.postgresql.org/docs/current/sql-prepare.html (grep for "Any SELECT, INSERT, UPDATE, DELETE, or VALUES statement.").

Also:
```
materialize=> PREPARE s AS CREATE ROLE foo;
ERROR:  unpreparable statement
LINE 1: PREPARE s AS CREATE ROLE foo;
                     ^
```

Maybe we should revisit this for privileges. There are some cases where we do bypass `sequence_plan` which might circumvent the RBAC checks:
- `CREATE SOURCE`.
- `CREATE ROLE` (this one is needed because we can't run RBAC checks when creating a role during startup).
- `COMMIT`
- `COPY ROWS`

Reply:
I reverted this for now and will think if/how to better handle it.

Reply:
So I think for this PR this is fine since only a few DDL statements are applicable for attribute checks and none of them are able to be prepared or executed.

After this PR I'll do the following:
- Put all `sequence_X` methods in a private module only accessible by `sequence_plan`.
- Modify every caller to only call `sequence_plan`.
- The one exception will be when creating a role during startup. I'll make a public method called `sequence_create_role_for_startup` that just calls out to `sequence_create_role` but makes it obvious that it should only be used startup.

This should increase our confidence that all statements will go through `sequence_plan` and not be able to circumvent the permissions check. Any thoughts?

Reply:
SGTM!